src/main/java/org/gradle/profiler/CommandExec.java

```diff
@@ -39,57 +42,171 @@ public String runAndCollectOutput(List<String> commandLine) {
+                    while (true) {
+                        try {
+                            int read = inputStream.read(buffer);
+                            output.write(buffer);
+                            if (read == -1) {
+                                output.flush();
+                                output.close();
+                                break;
+                            }
+                        } catch (IOException e) {
+                            throw new RuntimeException("Could not write input", e);
+                        }
+                    }
+                });
+            }
+            return new RunHandle(processBuilder, process, executor);
+        } catch (IOException e) {
+            throw new RuntimeException(commandErrorMessage(processBuilder), e);
         }
-        if (result != 0) {
-            throw new RuntimeException("Could not run command " + processBuilder.command().stream().collect(Collectors.joining(" ")));
+    }
+
+    private boolean readStream(InputStream inputStream, OutputStream outputStream, String command, byte[] buffer) {
+        int nread;
+        try {
+            nread = inputStream.read(buffer);
+        } catch (IOException e) {
+            throw new RuntimeException("Could not read input from child process for command '" + command + "'", e);
+        }
+        if (nread < 0) {
+            return true;
+        }
+        try {
+            outputStream.write(buffer, 0, nread);
+        } catch (IOException e) {
+            throw new RuntimeException("Could not write output from child process for command '" + command + "'", e);
+        }
+        return false;
+    }
+
+    private String commandErrorMessage(ProcessBuilder processBuilder) {
+        return "Could not run command " + processBuilder.command().stream().collect(Collectors.joining(" "));
+    }
+
+    public class RunHandle {
+        private final ProcessBuilder processBuilder;
+        private final Process process;
+        private final ExecutorService executor;
+
+        RunHandle(ProcessBuilder processBuilder, Process process, ExecutorService executor) {
+            this.processBuilder = processBuilder;
+            this.process = process;
+            this.executor = executor;
+        }
+
+        public void waitForSuccess() {
+            int result;
+            try {
+                result = process.waitFor();
+            } catch (Exception e) {
+                throw new RuntimeException(commandErrorMessage(processBuilder), e);
+            } finally {
+                shutdownExecutor();
+            }
+            if (result != 0) {
+                throw new RuntimeException(commandErrorMessage(processBuilder) + ". Exited with result " + result);
+            }
+        }
+
+        public void interrupt() {
+            int pid;
+            try {
+                Field field = process.getClass().getDeclaredField("pid");
```

Comment:
Hmm, not very useful I'm afraid - it just abstracts `destroy` for `JavaProcess` and I couldn't get `destroy` to send a sigint and let `perf` flush it's write buffer safely.